### PR TITLE
Bug 1244325 - adjust the pulse job YAML schema to accommodate task TC needs

### PR DESCRIPTION
**Pull request by camd**
These changes reflect updates to the ``pulse-job.yml`` file.  Some of these changes are based on the discussion here: https://public.etherpad-mozilla.org/p/jonasfj-teach-TH-retries-chunking

Summary of changes:

1. ``job_guid`` has been changed to a combination of ``taskId`` and ``retryId`` where the ``taskId`` *can* simply be the same as the old ``job_guid`` but the ``retryId`` will indicate how many times this job has been retried.  The values will be concatenated to <taskId>/<retryId> to form the ``job_guid`` value currently in the TH schema.

2. ``jobSymbol`` and ``chunkId`` will be concatenated to create the ``job_type_symbol`` internally.  But this allows us flexibility in the schema to separate the two in the future.

3. ``labels`` are being used in place of ``optionCollection`` to indicate more clearly that these are really just arbitrary strings that indicate attributes of the platform.  As we move forward to a wider matrix of platform configurations, the hope is that the term ``label`` is less confusing.  ``tag`` could have been used here, too.  But ``tag`` could possibly be confused with a tag in a version control system.  These values are, however, still translated to the ``option_collection`` internally to match the existing schema.

4. ``owner`` is used in place of ``who``.  Still maps to ``who`` internally for jobs.  But, again, the hope is that this term is a little clearer.

5. The general ``artifacts`` field has been removed and several ``artifacts`` are now top-level fields.  ``text_log_summary``, ``Bug suggestions``, and ``Job Info`` are not mere artifacts.  These are used heavily by the TH UI and should really become 1st class citizens if we continue to use them.  In this way, their format is now strictly defined.

6. An ``extra`` section has been added.  If the field of ``extra.artifacts`` is present, we will do our best to add those as arbitrary artifacts.  Anything else in the ``extra`` section is not yet used, but is flexible enough to be put to use later.

[Review on Reviewable](https://reviewable.io/reviews/mozilla/treeherder/1286)

### CODE REVIEW

File: schemas/pulse-job.yml [hunk 1]
```diff
@@ -116,9 +124,11 @@ properties:
       - unknown
   jobKind:
     type: "string"
+    default: "other"
     enum:
       - build
       - test
+      - other
   tier:
```
Comment (camd) [inline schemas/pulse-job.yml]: I'd rather stay with just an ``integer`` for now.  If we need a 2.5 or something, in the future, we can change it from ``integer`` to ``float`` (``number``) easy enough

File: schemas/pulse-job.yml [hunk 2]
```diff
@@ -170,9 +185,10 @@ properties:
       maxLength: 50
       pattern: "^[A-Za-z0-9_-]+$"
 
-  who:
-    title: "who"
-    format: "email"
+  owner:
```
Comment (camd) [inline schemas/pulse-job.yml]: as the description denotes, this won't always be an email.  so leaving free-form text for now.

File: schemas/pulse-job.yml [hunk 3]
```diff
@@ -170,9 +185,10 @@ properties:
...
+    description: |
+      Description of who submitted the job: gaia | scheduler name | username | email
+    title: "owner"
     type: "string"
     minLength: 1
     maxLength: 50
```
Comment (jonasfj) [inline schemas/pulse-job.yml]: So we're not going to have an email anymore?

Comment (jonasfj) [inline schemas/pulse-job.yml]: nvm, optional email...

File: schemas/pulse-job.yml (hunk 1 above)
Comment (jonasfj) [inline schemas/pulse-job.yml]: I'm okay with this... But warning you that when designing a  schema there is little "for now" :)
If you have 10 services feed into this, "for now" becomes "forever" very easily... Doesn't mean you can't change, just means you'll likely have to convert old value to new ones in your ingestion code.

Granted you can always change to `enum: [1,2,3 ,2.5]` that'll be backwards compatible. So this is fine.

File: schemas/pulse-job.yml [hunk 4]
```diff
@@ -55,13 +80,22 @@ properties:
...
     type: "object"
     properties:
-      jobSymbol:
-        title: "jobSymbol"
-        type: "string"
-        # spaces and "?" are not valid for job symbols
-        pattern: "^[A-Za-z0-9._-]+$"
-        minLength: 1
-        maxLength: 25
+      anyOf:
```
Comment (jonasfj) [inline schemas/pulse-job.yml]: `anyOf` doesn't work this way...

Wouldn't you want both `jobSymbol` and `chunkId` anyways?
It's just we can facilicate more than one with the same `jobSymbol` as long as `chunkId` is updated.

File: schemas/pulse-job.yml [hunk 5]
```diff
@@ -204,28 +241,180 @@ properties:
...
+                  format: date-time
+                error_count:
+                  type: integer
+                  minimum: 0
+                duration:
+                  type: integer
+                  minimum: 0
+                order:
+                  type: integer
+                  minimum: 0
```
Comment (jonasfj) [inline schemas/pulse-job.yml]: What is the `order`? if this is the order in which the step comes, I strongly advice you remove it and require people to define the `steps` array as an ordered array.

File: schemas/pulse-job.yml [hunk 6]
```diff
@@ -204,28 +241,180 @@ properties:
...
+                  items:
+                    type: object
+                    properties:
+                      line:
+                        type: string
+                        minLength: 1
+                        maxLength: 255
+                      linenumber:
+                        type: integer
+                        minimum: 0
```
Comment (jonasfj) [inline schemas/pulse-job.yml]: how about `line` and `lineNumber`

File: schemas/pulse-job.yml [hunk 7]
```diff
@@ -204,28 +241,180 @@ properties:
...
 
-  artifacts:
-    type: "array"
+  textLogSummary:
+    type: object
+    description: |
+      This object defines what is seen in the Treeherder Log Viewer.  This object
+      can be submitted here, or it will be generated from the submitted log.
+    properties:
+      step_data:
```
Comment (jonasfj) [inline schemas/pulse-job.yml]: too many nested objects, that don't have to be here... You can merge all properties of `step_data` into `textLogSummary`.

File: schemas/pulse-job.yml [hunk 8]
```diff
@@ -204,28 +241,180 @@ properties:
...
+                  minimum: 0
+                duration:
+                  type: integer
+                  minimum: 0
+                order:
+                  type: integer
+                  minimum: 0
+                result:
+                  type: string
+                  enum: [busted, testfailed, exception, success, canceled, unknown, retry, skipped]
```
Comment (jonasfj) [inline schemas/pulse-job.yml]: We normalized this at a higher level... should we do it here too?

File: schemas/pulse-job.yml [hunk 9]
```diff
@@ -204,28 +241,180 @@ properties:
   runMachine:
     $ref: "#/definitions/machine"
 
-  artifacts:
-    type: "array"
+  textLogSummary:
```
Comment (jonasfj) [inline schemas/pulse-job.yml]: In the interest of moving this forward I suggest what we leave a comment in the description advertizing `textLogSummary` as an experimental property.

I can't imagine we would actually want to support this format long-term. I'm not even sure it's sane to send this much data via. pulse... AMQP performance goes down if we send big messages.

File: schemas/pulse-job.yml [hunk 10]
```diff
@@ -204,28 +241,180 @@ properties:
...
-        - name
-        - blob
+          maxLength: 550
+        search_terms:
+          type: array
+          items:
+            type: string
+            minLength: 1
+            maxLength: 255
+        bugs:
```
Comment (jonasfj) [inline schemas/pulse-job.yml]: Unnecessary nesting...

Just make it `recentOpenBugs` and `otherBugs`...

File: schemas/pulse-job.yml [hunk 11]
```diff
@@ -204,28 +241,180 @@ properties:
...
+
+  jobInfo:
+    description: |
+      Definition of the Job Info for a job.  These are extra data
+      fields that go along with a job that will be displayed in
+      the details panel within Treeherder.
+    id: "jobInfo"
+    type: object
+    properties:
+      jobDetails:
```
Comment (jonasfj) [inline schemas/pulse-job.yml]: Unnecessary nesting...

File: schemas/pulse-job.yml [hunk 12]
```diff
@@ -204,28 +241,180 @@ properties:
...
+        items:
+        - type: object
+          properties:
+            url:
+              type: string
+              format: url
+            contentType:
+              type: string
+              enum:
+                - raw_html
```
Comment (jonasfj) [inline schemas/pulse-job.yml]: if you don't sanitize this it's an XSS security issue.

File: schemas/pulse-job.yml (hunk 9 above)
Comment (jonasfj) [inline schemas/pulse-job.yml]: I have similar concerns regarding `bugSuggestions` and `jobInfo`... but again... let's this working and then perfect it later...

I guess we can always mark the fields as "deprecated" when we have a better format...

**Comment by edmorley** [comment]
> job_guid has been changed to a combination of taskId and retryId where the taskId can simply be the same as the old job_guid but the retryId will indicate how many times this job has been retried. The values will be concatenated to / to form the job_guid value currently in the TH schema.

This seems very taskcluster specific, when I think we want this pulse schema to be generic to anyone. That said, job guids are something that we've talked about trying to get rid of, so not sure what the best approach is here.

**Comment by jonasfj** [comment]
>This seems very taskcluster specific, when I think we want this pulse schema to be generic to anyone. That said, job guids are something that we've talked about trying to get rid of, so not sure what the best approach is here.

Well, treeherder/sheriffs wants to know about infrastructure retries. In that case I want future treeherder to be one that can knows not just that "a job was retried", I should have the relational information and know what job it is retried in.

I previously proposed introducing `"retryOf": <other_job_guid>`, so a job could declare that it was a retry of a previous job. Hence, `<other_job_guid>` is declared retried. That would be equally okay to me.

But I don't like submitting a message saying "job <job_guid> is retried", TH needs to preserve the relationship between "old retried job", and "new job trying again".
Because then TH will never get to state where it only shows the job once and hides the retries by default.  And ideally infrastructure retries should not be something people concern themselves with :)

**Comment by edmorley** [comment]
Landed in https://github.com/mozilla/treeherder/commit/825913ece4e8bf653c58fac06e0ce7065a1ebb53

**Comment by edmorley** [comment]
(GitHub has this branch stuck and unable to delete; going to toggle to see if it fixes it)

**Pull request closed without merging**